rcl/src/rcl/parameter.c

```diff
@@ -0,0 +1,263 @@
+  const rcl_interfaces__msg__Parameter * src)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(dst, false);
+  RCL_CHECK_ARGUMENT_FOR_NULL(src, false);
+  if (!rosidl_generator_c__String__assign(&dst->name, src->name.data)) {
+    return RCL_RET_ERROR;
+  }
+  return rcl_parameter_value_copy(&dst->value, &src->value);
+}
+
+rcl_ret_t
+rcl_parameter_convert_changes_to_event(
+  const rcl_interfaces__msg__Parameter__Array * prior_state,
+  const rcl_interfaces__msg__Parameter__Array * new_state,
+  rcl_interfaces__msg__ParameterEvent * parameter_event)
+{
+  // Diff the prior state and the new state and fill the parameter_event struct
+  RCL_CHECK_ARGUMENT_FOR_NULL(prior_state, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(new_state, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(parameter_event, RCL_RET_INVALID_ARGUMENT);
+
+  size_t prior_idx, new_idx;
+  size_t num_deleted_params = 0;
+  size_t num_changed_params = 0;
+  size_t num_new_params = 0;
+  // go through and count the numbers we're going to need
+  for (prior_idx = 0; prior_idx < prior_state->size; ++prior_idx) {
+    rcl_interfaces__msg__Parameter * prior_entry = &prior_state->data[prior_idx];
+    for (new_idx = 0; new_idx < new_state->size; ++new_idx) {
+      rcl_interfaces__msg__Parameter * new_entry = &new_state->data[new_idx];
+      // If the parameter has the same name but a different type or value, count it as changed.
+      if (strcmp(prior_entry->name.data, new_entry->name.data) == 0) {
+        if (!rcl_parameter_value_compare(&prior_entry->value, &new_entry->value)) {
+          num_changed_params++;
+        }
+        break;
+      }
+    }
+    if (new_idx == new_state->size) {
+      // No entry in new_state was found with a matching name to prior_entry (deleted)
+      num_deleted_params++;
+    }
+  }
+  // Could optimize this if we allocated some space to store matches from the previous comparisons
+  for (new_idx = 0; new_idx < new_state->size; ++new_idx) {
+    rcl_interfaces__msg__Parameter * new_entry = &new_state->data[new_idx];
+    for (prior_idx = 0; prior_idx < prior_state->size; ++prior_idx) {
+      rcl_interfaces__msg__Parameter * prior_entry = &prior_state->data[prior_idx];
+      if (strcmp(prior_entry->name.data, new_entry->name.data) == 0) {
+        break;
+      }
+    }
+    if (prior_idx == prior_state->size) {
+      num_new_params++;
+    }
+  }
+
+  rcl_interfaces__msg__Parameter__Array__fini(&parameter_event->deleted_parameters);
+  rcl_interfaces__msg__Parameter__Array__fini(&parameter_event->changed_parameters);
+  rcl_interfaces__msg__Parameter__Array__fini(&parameter_event->new_parameters);
+
+  rcl_interfaces__msg__Parameter__Array__init(&parameter_event->deleted_parameters,
+    num_deleted_params);
+  rcl_interfaces__msg__Parameter__Array__init(&parameter_event->changed_parameters,
+    num_changed_params);
+  rcl_interfaces__msg__Parameter__Array__init(&parameter_event->new_parameters, num_new_params);
```

Comment:
Should we check for return values of these `rcl_interfaces__msg__Parameter__Array__init` calls in case memory cannot be allocated ?